keras_cv/custom_ops/kernels/withinbox_op.cc

```diff
@@ -0,0 +1,139 @@
+        return points_vec[a].y < points_vec[b].y;
+      });
+    std::vector<double> sorted_points_x;
+    sorted_points_x.reserve(num_points);
+    std::vector<double> sorted_points_y;
+    sorted_points_y.reserve(num_points);
+    for (int i = 0; i < num_points; ++i) {
+      sorted_points_x.emplace_back(points_vec[p_indices_x[i]].x);
+      sorted_points_y.emplace_back(points_vec[p_indices_y[i]].y);
+    }
+
+    // for each box, find all point indices whose x values are within box boundaries
+    // when the box is rotated, the box boundary is the minimum and maximum x for all vertices
+    std::vector<int> points_x_min = box::GetMinXIndexFromBoxes(boxes_vec, sorted_points_x);
+    std::vector<int> points_x_max = box::GetMaxXIndexFromBoxes(boxes_vec, sorted_points_x);
+    std::vector<std::unordered_set<int>> points_x_indices(num_boxes);
+    auto set_fn_x = [&points_x_min, &points_x_max, &p_indices_x, &points_x_indices](int64_t begin, int64_t end) {
+      for (int64_t idx = begin; idx < end; ++idx) {
+        std::unordered_set<int> p_set;
+        int p_start = points_x_min[idx];
+        int p_end = points_x_max[idx];
+        for (auto p_idx = p_start; p_idx <= p_end; ++p_idx) {
+          p_set.insert(p_indices_x[p_idx]);
+        }  
+        points_x_indices[idx] = p_set;
+      }
+    };
+    const CPUDevice& device = ctx->eigen_device<CPUDevice>();
+    const Eigen::TensorOpCost cost(num_points, num_boxes, 3);
+    device.parallelFor(num_boxes, cost, set_fn_x);
+
+    // for each box, find all point indices whose y values are within box boundaries
+    // when the box is rotated, the box boundary is the minimum and maximum x for all vertices
+    std::vector<int> points_y_min = box::GetMinYIndexFromBoxes(boxes_vec, sorted_points_y);
+    std::vector<int> points_y_max = box::GetMaxYIndexFromBoxes(boxes_vec, sorted_points_y);
+    std::vector<std::unordered_set<int>> points_y_indices(num_boxes);
+    auto set_fn_y = [&points_y_min, &points_y_max, &p_indices_y, &points_y_indices](int64_t begin, int64_t end) {
+      for (int64_t idx = begin; idx < end; ++idx) {
+        std::unordered_set<int> p_set;
+        int p_start = points_y_min[idx];
+        int p_end = points_y_max[idx];
+        for (auto p_idx = p_start; p_idx <= p_end; ++p_idx) {
+          p_set.insert(p_indices_y[p_idx]);
+        }  
+        points_y_indices[idx] = p_set;
+      }
+    };
+    device.parallelFor(num_boxes, cost, set_fn_y);
+
+    // for the intersection of x indices set and y indices set, check if
+    // those points are within the box
+    auto within_fn = [&points_x_indices, &points_y_indices, &boxes_vec, &points_vec, &boxes_indices_t](int64_t begin, int64_t end) {
+      for (int64_t idx = begin; idx < end; ++idx) {
+        std::unordered_set<int>& set_a = points_x_indices[idx];
+        std::unordered_set<int>& set_b = points_y_indices[idx];
+        std::unordered_set<int> p_set;
+        for (auto val : set_a) {
```

Comment:
Nit: std::set_intersection?

Reply:
that only works for sorted range of set_a and set_b